kore/src/Kore/Parser/Lexer.x

```diff
@@ -0,0 +1,314 @@
+token.
+-}
+tok_string :: AlexInput -> Int -> Alex Token
+tok_string AlexInput{alexPosn=p@(AlexPn fp _ line column), alexStr} len = do
+  let text = decodeUtf8 (ByteString.take len alexStr)
+      unescaped = unescape text in case unescaped of
+                                        Left str -> alexError fp line column str
+                                        Right t -> return $ Token p (TokenString t)
+   
+-- | Data type for Tokens. Contains filename, location info, and TokenClass
+data Token = Token AlexPosn TokenClass
+  deriving stock (Eq)
+  deriving stock (Show)
+
+{- | Get the Text argument of a Token whose TokenClass contains such an 
+argument
+-}
+getTokenBody :: Token -> Text
+getTokenBody (Token _ (TokenIdent t)) = t
+getTokenBody (Token _ (TokenSetIdent t)) = t
+getTokenBody (Token _ (TokenString t)) = t
+getTokenBody _ = error "getTokenBody can only be called on tokens which contain a Text field"
+
+
+-- | Get the TokenClass of a Token
+getTokenClass :: Token -> TokenClass
+getTokenClass (Token _ cls) = cls
+
+{- | Data type for the raw lexical data of a Token. Essentially an enumeration
+specifying which token a particular Token represents. Additionally contains the
+semantic data associated with identifier and string tokens.
+-}
+data TokenClass
+  = TokenModule
+  | TokenEndModule
+  | TokenImport
+  | TokenSort
+  | TokenSymbol
+  | TokenWhere
+  | TokenAlias
+  | TokenAxiom
+  | TokenClaim
+  | TokenHookedSort
+  | TokenHookedSymbol
+  | TokenColonEqual
+  | TokenColon
+  | TokenLeftBrace
+  | TokenRightBrace
+  | TokenLeftBracket
+  | TokenRightBracket
+  | TokenLeftParen
+  | TokenRightParen
+  | TokenComma
+  | TokenTop
+  | TokenBottom
+  | TokenNot
+  | TokenAnd
+  | TokenOr
+  | TokenImplies
+  | TokenIff
+  | TokenExists
+  | TokenForall
+  | TokenMu
+  | TokenNu
+  | TokenCeil
+  | TokenFloor
+  | TokenEquals
+  | TokenIn
+  | TokenNext
+  | TokenRewrites
+  | TokenDomainValue
+  | TokenLeftAssoc
+  | TokenRightAssoc
+  | TokenIdent Text
+  | TokenSetIdent Text
+  | TokenString Text
+  | TokenEOF
+  deriving stock (Eq, Show)
+
+
+alexScanUser :: () -> AlexInput -> Int -> AlexReturn (AlexInput -> Int -> Alex Token)
+
+data AlexInput = AlexInput { alexPosn :: {-# UNPACK #-} !AlexPosn,
+                             alexChar :: {-# UNPACK #-} !Char,
+                             alexStr :: {-# UNPACK #-} !ByteString.ByteString,
+                             alexBytePos :: {-# UNPACK #-} !Int}
+
+data AlexPosn = AlexPn !FilePath !Int !Int !Int
+        deriving (Eq,Show)
+
+data AlexState = AlexState {
+        alex_pos :: !AlexPosn,  -- position at current input location
+        alex_bpos:: !Int,     -- bytes consumed so far
+        alex_inp :: ByteString.ByteString,      -- the current input
+        alex_chr :: !Char,      -- the character before the input
+        alex_scd :: !Int        -- the current startcode
+    }
+
+newtype Alex a = Alex { unAlex :: AlexState -> Either String (AlexState, a) }
```

Comment:
Essentially, yes. I tinkered with using that implementation and a deriving via declaration, but it turned out to be slower than this code, so I left it like this. This is also code that was copied from the Alex wrapper.

Reply:
Hmm. I think I told you it was `ExceptT String (State AlexState)` but in fact Ana is correct - it's the other way around. These transformers do not commute with each other and it probably makes a performance difference.

Reply:
I mean, the code that I tried out was `StateT AlexState (Either String) a` because what you suggested did not compile so I assumed that you had made that mistake. But as I said, it was slower than this code.